File: backend/shared/core/src/main/java/io/mateu/core/application/mcp/McpProjection.java
package io.mateu.core.application.mcp;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ArrayNode;
import com.fasterxml.jackson.databind.node.ObjectNode;
import java.util.LinkedHashMap;
import java.util.Map;

/**
 * Projects a serialized {@code UIIncrementDto} (the wire) into a flat, agent-friendly screen.
 *
 * <p>This is the native (Java) twin of the sidecar's {@code projection.mjs}. Both walk the SAME
 * serialized wire, so they produce the SAME projection — parity by construction. The derivation
 * rules are normative and specified in {@code doc/.../reference/wire-specification.md} (§ Agent
 * operability). Part of the agent-operability plane ({@code design/riu-agent-operability-plan.md}).
 *
 * <p>Pure: no I/O, so it is fully unit-testable against the conformance corpus.
 */
public final class McpProjection {

  private McpProjection() {}

  /** The server encodes the root/empty route as "_empty" on the wire; treat it as "". */
  public static String normalizeRoute(String route) {
    if (route == null) return "";
    String r = route.replaceFirst("^/+", "");
    return "_empty".equals(r) ? "" : r;
  }

  public static ObjectNode project(JsonNode increment, ObjectMapper om) {
    ObjectNode out = om.createObjectNode();
    JsonNode inc = increment == null ? om.createObjectNode() : increment;

    JsonNode commands = inc.path("commands");
    JsonNode messages = inc.path("messages");
    JsonNode fragments = inc.path("fragments");

    JsonNode[] serverSide = {null};
    JsonNode[] page = {null};
    JsonNode[] crudl = {null};
    java.util.List<JsonNode> fieldNodes = new java.util.ArrayList<>();
    Map<String, String> buttonsByAction = new LinkedHashMap<>();
    JsonNode[] state = {null};

    if (fragments.isArray()) {
      for (JsonNode fragment : fragments) {
        if (state[0] == null && fragment.path("state").isObject()) state[0] = fragment.get("state");
        JsonNode root = fragment.has("component") ? fragment.get("component") : fragment;
        deepVisit(
            root,
            node -> {
              if ("ServerSide".equals(node.path("type").asText(null)) && serverSide[0] == null) {
                serverSide[0] = node;
              }
              String t = mdType(node);
              if ("FormField".equals(t) && node.path("metadata").hasNonNull("fieldId")) {
                fieldNodes.add(node.get("metadata"));
              } else if ("Page".equals(t) && page[0] == null) {
                page[0] = node.get("metadata");
              } else if ("Crudl".equals(t) && crudl[0] == null) {
                crudl[0] = node.get("metadata");
              } else if ("Button".equals(t) && node.path("metadata").hasNonNull("actionId")) {
                String actionId = node.get("metadata").get("actionId").asText();
                buttonsByAction.putIfAbsent(
                    actionId, node.get("metadata").path("label").asText(null));
              }
            });
      }
    }

    JsonNode values =
        state[0] != null
            ? state[0]
            : (serverSide[0] != null && serverSide[0].path("initialData").isObject()
                ? serverSide[0].get("initialData")
                : om.createObjectNode());

    // Fields (deduped by id, declaration order).
    ArrayNode fields = out.putArray("fields");
    java.util.Set<String> seenField = new java.util.HashSet<>();
    for (JsonNode md : fieldNodes) {
      String id = md.get("fieldId").asText();
      if (!seenField.add(id)) continue;
      ObjectNode f = fields.addObject();
      f.put("id", id);
      f.put("label", md.path("label").asText(id));
      f.put("dataType", md.path("dataType").asText("string"));
      f.put("stereotype", md.path("stereotype").asText("regular"));
      f.put("required", md.path("required").asBoolean(false));
      f.put("readOnly", md.path("readOnly").asBoolean(false));
      if (values.has(id)) f.set("value", values.get(id));
      if (md.path("options").isArray() && md.get("options").size() > 0) {
        ArrayNode opts = f.putArray("options");
        for (JsonNode o : md.get("options")) {
          ObjectNode oo = opts.addObject();
          oo.set("value", o.get("value"));
          oo.put("label", o.path("label").asText(o.path("value").asText("")));
        }
      }
    }

    // Actions: ids the component claims (RBAC already applied server-side), labels from buttons.
    ArrayNode actions = out.putArray("actions");
    java.util.Set<String> seenAction = new java.util.HashSet<>();
    if (serverSide[0] != null && serverSide[0].path("actions").isArray()) {
      for (JsonNode a : serverSide[0].get("actions")) {
        String id = a.path("id").asText(null);
        if (id == null || !seenAction.add(id)) continue;
        ObjectNode ao = actions.addObject();
        ao.put("id", id);
        ao.put(
            "label",
            buttonsByAction.getOrDefault(id, id) != null
                ? buttonsByAction.getOrDefault(id, id)
                : id);
        if (a.hasNonNull("shortcut") && !a.get("shortcut").asText().isEmpty())
          ao.put("shortcut", a.get("shortcut").asText());
        if (a.path("confirmationRequired").asBoolean(false)) ao.put("confirmationRequired", true);
        if (a.hasNonNull("href") && !a.get("href").asText().isEmpty())
          ao.put("href", a.get("href").asText());
      }
    }
    for (Map.Entry<String, String> e : buttonsByAction.entrySet()) {
      if (seenAction.add(e.getKey())) {
        ObjectNode ao = actions.addObject();
        ao.put("id", e.getKey());
        ao.put("label", e.getValue() != null ? e.getValue() : e.getKey());
      }
    }

    // Title / subtitle.
    String title = commandData(commands, "SetWindowTitle");
    if (title == null && page[0] != null) {
      title =
          page[0].hasNonNull("pageTitle")
              ? page[0].get("pageTitle").asText()
              : page[0].path("title").asText(null);
    }
    if (title == null && crudl[0] != null) title = crudl[0].path("title").asText(null);

    out.put(
        "route",
        serverSide[0] != null ? normalizeRoute(serverSide[0].path("route").asText(null)) : null);
    out.put(
        "serverSideType",
        serverSide[0] != null ? serverSide[0].path("serverSideType").asText(null) : null);
    String pageType =
        serverSide[0] != null && serverSide[0].hasNonNull("pageType")
            ? serverSide[0].get("pageType").asText()
            : (page[0] != null ? page[0].path("pageType").asText(null) : null);
    out.put("pageType", pageType);
    out.put("wireVersion", inc.path("wireVersion").asText(null));
    out.put("title", title);
    out.put(
        "subtitle",
        page[0] != null && page[0].hasNonNull("subtitle")
            ? page[0].get("subtitle").asText()
            : (crudl[0] != null ? crudl[0].path("subtitle").asText(null) : null));
    out.set("state", values);

    if (crudl[0] != null) {
      ObjectNode listing = out.putObject("listing");
      listing.put("title", crudl[0].path("title").asText(null));
      listing.put("searchable", crudl[0].path("searchable").asBoolean(false));
      ArrayNode columns = listing.putArray("columns");
      if (crudl[0].path("columns").isArray()) {
        for (JsonNode c : crudl[0].get("columns")) {
          JsonNode md = c.has("metadata") ? c.get("metadata") : c;
          String id = md.hasNonNull("id") ? md.get("id").asText() : md.path("fieldId").asText(null);
          String label =
              md.hasNonNull("caption") ? md.get("caption").asText() : md.path("label").asText(null);
          if (id != null || label != null) {
            ObjectNode co = columns.addObject();
            co.put("id", id);
            co.put("label", label);
          }
        }
      }
      ArrayNode filters = listing.putArray("filters");
      if (crudl[0].path("filters").isArray()) {
        for (JsonNode fnode : crudl[0].get("filters")) {
          ObjectNode fo = filters.addObject();
          fo.put("id", fnode.path("fieldId").asText(null));
          fo.put("label", fnode.path("label").asText(fnode.path("fieldId").asText(null)));
          fo.put("dataType", fnode.path("dataType").asText("string"));
        }
      }
    }

    if (messages.isArray() && messages.size() > 0) {
      ArrayNode ms = out.putArray("messages");
      for (JsonNode m : messages) {
        ObjectNode mo = ms.addObject();
        mo.put(
            "text", m.hasNonNull("text") ? m.get("text").asText() : m.path("message").asText(""));
        mo.put("type", m.path("type").asText("info"));
      }
    }
    if (commands.isArray()) {
      ArrayNode cs = om.createArrayNode();
      for (JsonNode c : commands) {
        String type = c.path("type").asText(null);
        if (type != null && !"SetWindowTitle".equals(type)) {
          ObjectNode co = cs.addObject();
          co.put("type", type);
          co.set("data", c.get("data"));
        }
      }
      if (cs.size() > 0) out.set("commands", cs);
    }

    return out;
  }

  private static String commandData(JsonNode commands, String type) {
    if (!commands.isArray()) return null;
    for (JsonNode c : commands) {
      if (type.equals(c.path("type").asText(null))) {
        return c.hasNonNull("data") ? c.get("data").asText() : null;
      }
    }
    return null;
  }

  private static String mdType(JsonNode node) {
    JsonNode md = node.path("metadata");
    return md.isObject() && md.hasNonNull("type") ? md.get("type").asText() : null;
  }

  private interface NodeVisitor {
    void visit(JsonNode node);
  }

  private static void deepVisit(JsonNode node, NodeVisitor fn) {
    if (node == null || node.isNull()) return;
    if (node.isObject()) {
      fn.visit(node);
      for (JsonNode child : node) deepVisit(child, fn);
    } else if (node.isArray()) {
      for (JsonNode child : node) deepVisit(child, fn);
    }
  }
}

File: backend/dotnet/src/Mateu.Core/Mcp/McpProjection.cs
using System.Text.Json.Nodes;

namespace Mateu.Core.Mcp;

/// <summary>
/// Projects a serialized UIIncrementDto (the wire) into a flat, agent-friendly screen — the C#
/// twin of the sidecar's projection.mjs and the Java McpProjection. All hosts walk the SAME
/// serialized wire, so they produce the SAME projection (parity by construction). The derivation
/// rules are normative (doc/.../reference/wire-specification.md § Agent operability). Part of the
/// agent-operability plane (design/riu-agent-operability-plan.md, P4).
///
/// <para>Pure (no I/O). A System.Text.Json.Nodes node has a single parent, so any node copied from
/// the source into the output is <c>DeepClone()</c>d.</para>
/// </summary>
public static class McpProjection
{
    /// <summary>The server encodes the root/empty route as "_empty"; treat it as "".</summary>
    public static string NormalizeRoute(string? route)
    {
        if (route is null) return "";
        var r = route.TrimStart('/');
        return r == "_empty" ? "" : r;
    }

    public static JsonObject Project(JsonNode? increment)
    {
        var inc = increment as JsonObject ?? new JsonObject();
        var commands = inc["commands"] as JsonArray ?? new JsonArray();
        var messages = inc["messages"] as JsonArray ?? new JsonArray();
        var fragments = inc["fragments"] as JsonArray ?? new JsonArray();

        JsonObject? serverSide = null, page = null, crudl = null, state = null;
        var fieldNodes = new List<JsonObject>();
        var buttonsByAction = new List<KeyValuePair<string, string?>>();
        var buttonSeen = new HashSet<string>();

        foreach (var fragEl in fragments)
        {
            if (fragEl is not JsonObject fragment) continue;
            if (state is null && fragment["state"] is JsonObject st) state = st;
            var root = fragment["component"] ?? fragment;
            DeepVisit(root, node =>
            {
                if (Str(node["type"]) == "ServerSide" && serverSide is null) serverSide = node;
                var t = MdType(node);
                var md = node["metadata"] as JsonObject;
                if (t == "FormField" && md is { } mf && mf["fieldId"] is not null) fieldNodes.Add(mf);
                else if (t == "Page" && page is null) page = md;
                else if (t == "Crudl" && crudl is null) crudl = md;
                else if (t == "Button" && md is { } mb && mb["actionId"] is not null)
                {
                    var aid = Str(mb["actionId"])!;
                    if (buttonSeen.Add(aid)) buttonsByAction.Add(new(aid, Str(mb["label"])));
                }
            });
        }

        var values = state ?? serverSide?["initialData"] as JsonObject ?? new JsonObject();

        // Fields (deduped by id, declaration order).
        var fields = new JsonArray();
        var seenField = new HashSet<string>();
        foreach (var md in fieldNodes)
        {
            var fid = Str(md["fieldId"])!;
            if (!seenField.Add(fid)) continue;
            var f = new JsonObject
            {
                ["id"] = fid,
                ["label"] = Str(md["label"]) ?? fid,
                ["dataType"] = Str(md["dataType"]) ?? "string",
                ["stereotype"] = Str(md["stereotype"]) ?? "regular",
                ["required"] = Bool(md["required"]),
                ["readOnly"] = Bool(md["readOnly"]),
            };
            if (values.ContainsKey(fid)) f["value"] = values[fid]?.DeepClone();
            if (md["options"] is JsonArray opts && opts.Count > 0)
            {
                var arr = new JsonArray();
                foreach (var oEl in opts)
                {
                    if (oEl is not JsonObject o) continue;
                    arr.Add(new JsonObject
                    {
                        ["value"] = o["value"]?.DeepClone(),
                        ["label"] = Str(o["label"]) ?? o["value"]?.ToString() ?? "",
                    });
                }
                f["options"] = arr;
            }
            fields.Add(f);
        }

        // Actions: ids the component claims (RBAC already applied server-side), labels from buttons.
        var actions = new JsonArray();
        var seenAction = new HashSet<string>();
        if (serverSide?["actions"] is JsonArray declared)
        {
            foreach (var aEl in declared)
            {
                if (aEl is not JsonObject a) continue;
                var id = Str(a["id"]);
                if (id is null || !seenAction.Add(id)) continue;
                var ao = new JsonObject { ["id"] = id, ["label"] = Label(buttonsByAction, id) ?? id };
                var shortcut = Str(a["shortcut"]);
                if (!string.IsNullOrEmpty(shortcut)) ao["shortcut"] = shortcut;
                if (Bool(a["confirmationRequired"])) ao["confirmationRequired"] = true;
                var href = Str(a["href"]);
                if (!string.IsNullOrEmpty(href)) ao["href"] = href;
                actions.Add(ao);
            }
        }
        foreach (var kv in buttonsByAction)
        {
            if (seenAction.Add(kv.Key))
                actions.Add(new JsonObject { ["id"] = kv.Key, ["label"] = kv.Value ?? kv.Key });
        }

        string? title = CommandData(commands, "SetWindowTitle");
        if (title is null && page is not null) title = Str(page["pageTitle"]) ?? Str(page["title"]);
        if (title is null && crudl is not null) title = Str(crudl["title"]);

        var outObj = new JsonObject
        {
            ["route"] = serverSide is not null ? NormalizeRoute(Str(serverSide["route"])) : null,
            ["serverSideType"] = serverSide is not null ? Str(serverSide["serverSideType"]) : null,
            ["pageType"] = (serverSide is not null ? Str(serverSide["pageType"]) : null)
                           ?? (page is not null ? Str(page["pageType"]) : null),
            ["wireVersion"] = Str(inc["wireVersion"]),
            ["title"] = title,
            ["subtitle"] = (page is not null ? Str(page["subtitle"]) : null)
                           ?? (crudl is not null ? Str(crudl["subtitle"]) : null),
            ["fields"] = fields,
            ["actions"] = actions,
            ["state"] = values.DeepClone(),
        };

        if (crudl is not null)
        {
            var columns = new JsonArray();
            if (crudl["columns"] is JsonArray cols)
                foreach (var cEl in cols)
                {
                    var md = (cEl as JsonObject)?["metadata"] as JsonObject ?? cEl as JsonObject;
                    var cid = Str(md?["id"]) ?? Str(md?["fieldId"]);
                    var label = Str(md?["caption"]) ?? Str(md?["label"]);
                    if (cid is not null || label is not null)
                        columns.Add(new JsonObject { ["id"] = cid, ["label"] = label });
                }
            var filters = new JsonArray();
            if (crudl["filters"] is JsonArray fs)
                foreach (var fEl in fs)
                {
                    if (fEl is not JsonObject fo) continue;
                    filters.Add(new JsonObject
                    {
                        ["id"] = Str(fo["fieldId"]),
                        ["label"] = Str(fo["label"]) ?? Str(fo["fieldId"]),
                        ["dataType"] = Str(fo["dataType"]) ?? "string",
                    });
                }
            outObj["listing"] = new JsonObject
            {
                ["title"] = Str(crudl["title"]),
                ["searchable"] = Bool(crudl["searchable"]),
                ["columns"] = columns,
                ["filters"] = filters,
            };
        }

        if (messages.Count > 0)
        {
            var ms = new JsonArray();
            foreach (var mEl in messages)
            {
                if (mEl is not JsonObject m) continue;
                ms.Add(new JsonObject
                {
                    ["text"] = Str(m["text"]) ?? Str(m["message"]) ?? "",
                    ["type"] = Str(m["type"]) ?? "info",
                });
            }
            outObj["messages"] = ms;
        }

        var notable = new JsonArray();
        foreach (var cEl in commands)
        {
            if (cEl is not JsonObject c) continue;
            var type = Str(c["type"]);
            if (type is not null && type != "SetWindowTitle")
                notable.Add(new JsonObject { ["type"] = type, ["data"] = c["data"]?.DeepClone() });
        }
        if (notable.Count > 0) outObj["commands"] = notable;

        return outObj;
    }

    static string? Label(List<KeyValuePair<string, string?>> buttons, string id)
    {
        foreach (var kv in buttons)
            if (kv.Key == id) return kv.Value;
        return null;
    }

    static string? CommandData(JsonArray commands, string type)
    {
        foreach (var cEl in commands)
            if (cEl is JsonObject c && Str(c["type"]) == type) return Str(c["data"]);
        return null;
    }

    static string? MdType(JsonObject node) =>
        node["metadata"] is JsonObject md ? Str(md["type"]) : null;

    static void DeepVisit(JsonNode? node, Action<JsonObject> fn)
    {
        if (node is JsonObject obj)
        {
            fn(obj);
            foreach (var kv in obj) DeepVisit(kv.Value, fn);
        }
        else if (node is JsonArray arr)
        {
            foreach (var item in arr) DeepVisit(item, fn);
        }
    }

    static string? Str(JsonNode? n) => n is JsonValue v && v.TryGetValue<string>(out var s) ? s : null;

    static bool Bool(JsonNode? n) => n is JsonValue v && v.TryGetValue<bool>(out var b) && b;
}
